pkg/operator2/datasync/syncdata.go

```diff
@@ -0,0 +1,238 @@
+			}
+		} else if cmErrs := validateConfigMap(cmLister, src); len(cmErrs) > 0 {
+			errs = append(errs, fmt.Errorf("error validating configMap openshift-config/%s: %w", src.Src, errors.NewAggregate(cmErrs)))
+		}
+	}
+	return errs
+}
+
+// AddIDPSecret initializes a sourceData object with proper data for a Secret
+// and adds it among the other secrets stored here
+// Returns the path for the Secret
+func (sd *ConfigSyncData) AddIDPSecret(index int, secretRef configv1.SecretNameReference, field, key string) string {
+	if len(secretRef.Name) == 0 {
+		return ""
+	}
+
+	dest, data := newSourceDataIDP(index, SecretType, secretRef.Name, field, key)
+	sd.data[dest] = data
+
+	return path.Join(data.Path, key)
+}
+
+// AddIDPConfigMap initializes a sourceData object with proper data for a ConfigMap
+// and adds it among the other configmaps stored here
+// Returns the path for the ConfigMap
+func (sd *ConfigSyncData) AddIDPConfigMap(index int, configMapRef configv1.ConfigMapNameReference, field, key string) string {
+	if len(configMapRef.Name) == 0 {
+		return ""
+	}
+
+	dest, data := newSourceDataIDP(index, ConfigMapType, configMapRef.Name, field, key)
+	sd.data[dest] = data
+
+	return path.Join(data.Path, key)
+}
+
+// ToVolumesAndMounts converts the synchronization data to Volumes and VoulumeMounts
+// so that these can be added to a container spec
+func (sd *ConfigSyncData) ToVolumesAndMounts() ([]corev1.Volume, []corev1.VolumeMount) {
+	volumes := []corev1.Volume{}
+	volumeMounts := []corev1.VolumeMount{}
+
+	// maps' keys are random,  we need to sort the output to prevent redeployment hotloops
+	for _, dataKey := range sets.StringKeySet(sd.data).List() {
+		volume, volumeMount := sd.data[dataKey].ToVolumesAndMounts(dataKey)
+		volumes = append(volumes, volume)
+		volumeMounts = append(volumeMounts, volumeMount)
+	}
+
+	return volumes, volumeMounts
+
+}
+
+func (s sourceData) ToVolumesAndMounts(volName string) (corev1.Volume, corev1.VolumeMount) {
+	vol := corev1.Volume{
+		Name: volName,
+	}
+
+	items := []corev1.KeyToPath{
+		{
+			Key:  s.Key,
+			Path: s.Key,
+		},
+	}
+
+	switch s.Resource {
+	case ConfigMapType:
+		vol.ConfigMap = &corev1.ConfigMapVolumeSource{
+			LocalObjectReference: corev1.LocalObjectReference{
+				Name: volName,
+			},
+			Items: items,
+		}
+	default:
+		vol.Secret = &corev1.SecretVolumeSource{
+			SecretName: volName,
+			Items:      items,
+		}
+	}
+
+	return vol, corev1.VolumeMount{
+		Name:      volName,
+		ReadOnly:  true,
+		MountPath: s.Path,
+	}
+}
+
+func getIDPName(i int, field string) string {
+	// idps that are synced have this prefix
+	return fmt.Sprintf("v4-0-config-user-idp-%d-%s", i, field)
+}
+
+func getIDPPath(i int, resource, dest string) string {
+	// root path for IDP data
+	return fmt.Sprintf("/var/config/user/idp/%d/%s/%s", i, resource, dest)
+}
+
+func syncOrDie(syncFunc func(dest, src resourcesynccontroller.ResourceLocation) error, dest, src string) {
```

Comment:
nit: Maybe factor out the common functionality between this and `syncSecretOrDie`?

Reply:
done, exported the sync function and reused